src/Microsoft.Extensions.Diagnostics.HealthChecks/HealthCheckService.cs

```diff
@@ -6,127 +6,103 @@
-            var names = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
-            var duplicates = new List<string>();
-            foreach (var check in healthChecks)
+            // We're specifically going out of our way to do this at startup time. We want to make sure you
+            // get any kind of health-check related error as early as possible. Waiting until someone
+            // actually tries to **run** health checks would be real baaaaad.
+            using (var scope = _scopeFactory.CreateScope())
             {
-                if (!names.Add(check.Name))
-                {
-                    duplicates.Add(check.Name);
-                }
+                var healthChecks = scope.ServiceProvider.GetRequiredService<IEnumerable<IHealthCheck>>();
+                EnsureNoDuplicates(healthChecks);
             }
+        }
 
-            if (duplicates.Count > 0)
+        public Task<CompositeHealthCheckResult> CheckHealthAsync(CancellationToken cancellationToken = default) =>
+            CheckHealthAsync(predicate: null, cancellationToken);
+
+        public async Task<CompositeHealthCheckResult> CheckHealthAsync(
+            Func<IHealthCheck, bool> predicate,
+            CancellationToken cancellationToken = default)
+        {
+            using (var scope = _scopeFactory.CreateScope())
             {
-                throw new ArgumentException($"Duplicate health checks were registered with the name(s): {string.Join(", ", duplicates)}", nameof(healthChecks));
-            }
+                var healthChecks = scope.ServiceProvider.GetRequiredService<IEnumerable<IHealthCheck>>();
 
-            Checks = healthChecks.ToDictionary(c => c.Name, StringComparer.OrdinalIgnoreCase);
+                if (predicate != null)
+                {
+                    healthChecks = healthChecks.Where(predicate);
+                }
 
-            if (_logger.IsEnabled(LogLevel.Debug))
-            {
-                foreach (var check in Checks)
+                var results = new Dictionary<string, HealthCheckResult>(StringComparer.OrdinalIgnoreCase);
+                foreach (var healthCheck in healthChecks)
                 {
-                    _logger.LogDebug("Health check '{healthCheckName}' has been registered", check.Key);
+                    cancellationToken.ThrowIfCancellationRequested();
+
+                    // If the health check does things like make Database queries using EF or backend HTTP calls,
+                    // it may be valuable to know that logs it generates are part of a health check. So we start a scope.
+                    using (_logger.BeginScope(new HealthCheckLogScope(healthCheck.Name)))
+                    {
+                        HealthCheckResult result;
+                        try
+                        {
+                            _logger.LogTrace("Running health check: {healthCheckName}", healthCheck.Name);
```

Comment:
This isn't the logging pattern we use in components.

Reply:
YEAH BRO